### ISSUE TITLE
selection on datashaded points plot using 'lasso_select' or 'poly_select' returns points that are not in the plot

### ISSUE THREAD

**Issue by hali-geoviz** [holoviz/holoviews#5086]
#### ALL software version info
holoviews: version: 1.14.5
param: version: 1.10.1
bokeh: version: 2.3.2
panel: version: 0.11.1
datashader: version: 0.12.1
spatialpandas: version: 0.4.3
shapely: version: 1.7.1
python: Python 3.7.10
Running inside Jupyter notebook
OS: Windows 10
browser: Chrome

#### When holoviews.operation.datashader.datashade is applied to a holoviews points plot and this is wrapped inside a holoviews link_selection the expected behaviour is that tools like box_select, lasso_select, and poly_select will return only rows where both of the kdims are real numbers. This seems to be correct for box_select, but not for the two others. I have experienced this to bee a bit unstable but am able to reproduce in the below example. 

This is an amazingly effective tool, but I am concerned with there being points selected that are not really inside the points plot. 

It could be my way of selecting the data from the holoviews dataset, an if so please close this issue. 

#### Complete, minimal, self-contained example code that reproduces the issue

```
# Imports
%load_ext autoreload
%autoreload 2

import numpy as np
import pandas as pd
import hvplot.pandas

# Import plotting libs
import holoviews as hv
from holoviews.operation.datashader import datashade
from holoviews.selection import link_selections
from holoviews.streams import PolyDraw
from holoviews import opts

# set bakend
hv.extension('bokeh')

# Prepare data

np.random.seed(1)
n = 10_000_000
n1 = n - 3_000_000
n2 = n - n1
ids = n1 * ['ONE'] +  n2 * ['TWO']

data = np.random.multivariate_normal((3,10), [[0.1, 0.1], [0.1, 1.0]], (n,))

SENS_A = data[:,0]
SENS_B = data[:,1]

# Make holes in data
SENS_A = np.where(np.random.rand(n) > .5, np.NaN, SENS_A)
SENS_B = np.where(np.random.rand(n) < .2, np.NaN, SENS_B)

data_dict = {'id': ids,
             'SENS_A': SENS_A,
             'SENS_B': SENS_B}

df = pd.DataFrame(data_dict)

# Ensure there are nan's  
try: 
    display(df.isna().sum())
    display(df.head())
except:
    pass

# Create plot
p = hv.Points(df, kdims=['SENS_A', 'SENS_B'])

ls = link_selections.instance(selection_mode='overwrite',
                              unselected_alpha=.4)
pp = datashade(p, cmap='fire').opts(height=600, width=600, tools=['lasso_select', 'poly_select', 'box_select'], active_tools=['lasso_select'])
ls(pp).opts(bgcolor='black')

# So some interactive selection with the tools 'box_select', 'lasso_select' and 'poly_select'
# Then apply the following:
df.loc[hv.Dataset(df).select(ls.selection_expr).data.index, ['SENS_A', 'SENS_B']]

# Then run this to ensure there are no points selected that are not having both kdims with real numbers. 
n_selected_rows_with_any_nan = df.loc[hv.Dataset(df).select(ls.selection_expr).data.index, ['SENS_A', 'SENS_B']].isna().any(axis=1).sum()
criterium = (n_selected_rows_with_any_nan == 0)
assert_msg = "Selected points exist in 2d space and should by definition not have any nan's"

assert criterium, assert_msg

# When using 'lasso_select' and 'poly_select' the returned result set seems to also contain rows where either or both kdims has NaN values. 

```

#### Screenshots or screencasts of the bug in action
![spatial_select example](https://user-images.githubusercontent.com/47379702/133753811-c4396594-fc70-4729-87be-d6d91d20a21a.gif)

I also have a post in discourse that was not picked up on yet. But there my 'poly_select' actually worked correctly. see [discourse link](https://discourse.holoviz.org/t/holoviews-spatialpandas-lasso-or-polygon-draw-select-on-shaded-points-plot/2774)

I wonder if this has to do with the presence of or combination of spatialpandas and shapely(?)

**Comment by jbednar** [comment]
Thanks for the report! I've updated the title to what I _think_ is the specific issue you are reporting, i.e. that when you use datashader + spatialpandas in a lasso or poly selection, values with NaN coordinates are returned and they shouldn't be. I haven't' checked the code, but my guess is that we need to add special handling for NaNs in the selection logic in spatialpandas; the current logic is presumably incorrectly matching NaN values when it is filtering.

**Comment by GeoVizNow** [comment]
Yes, thank you for updating to a more precise title! It would be very welcome with any better way (than I used) to fetch selections! 😊 This concept provides a really interesting way to be able to see through big datasets. Please notify me if I can in any way help testing any findings. What I think to have noticed is that shapely does the selection better(?)/differently, but much more slow.

**Comment by hali-geoviz** [comment]
Just providing a small update based on what I have just seen: 

# Running the below line after a selection indicates it may be a lot of erronous selections (with my current setup) 
`df.loc[hv.Dataset(df).select(ls.selection_expr).data.index, ['SENS_A', 'SENS_B']].isna().any(1).sum()
# Comparing to selections where both coordinates are numeric values it indicates the problem could have a not minor impact on the dataset if say any cleaning is done based on this. 
`df.loc[hv.Dataset(df).select(ls.selection_expr).data.index, ['SENS_A', 'SENS_B']].notna().all(1).sum()

Anyhow based on the test data provided by the code above it becomes evident that the majority of erroneous data are from the x-dimension. Running the below code indicates the histogram in y-dimension matches the selection expression well while the histogram in the x-dimension returns a lot of values outside the range of the selection expression. 
`df_sel = df.loc[hv.Dataset(df).select(ls.selection_expr).data.index, ['SENS_A', 'SENS_B']]

`df_sel['SENS_A'].hvplot.hist()

`df_sel['SENS_B'].hvplot.hist()

This might give a clue to where we might search for the cause.

**Comment by hali-geoviz** [comment]
Actually another clue is the following code which indicates that included in the selection is all rows where the y-dimension got NaN's. This result is when I select with lasso in a part of the plot where I cannot see any points really (lower right part of the plot) 

![image](https://user-images.githubusercontent.com/47379702/133943524-3326f0cc-9b61-4307-a4ea-d590c3efc687.png)

`df_sel['SENS_A'].hvplot.hist() + df.loc[df['SENS_B'].isna()]['SENS_A'].hvplot.hist() + df['SENS_A'].hvplot.hist() `

![image](https://user-images.githubusercontent.com/47379702/133943571-f78854c8-91cf-45e1-8310-7a78703e76a1.png)

**Comment by jlstevens** [comment]
From the discussion above, it sounds like a bug in how the spatialpandas support handles filtering.
